### PR TITLE
feat: implement toHaveSelection

### PR DESCRIPTION
**Pull request by silviuaavram**
**What**:
Implement `toHaveSelection`.

**Why**:
Closes https://github.com/testing-library/jest-dom/issues/289.

**How**:
Based on https://github.com/testing-library/jest-dom/pull/412.

Differences from original PR:
- changed the partial selection condition a little bit to help with testing:
```js
else {
    // Element is partially selected
    const selectionStartsWithinElement =
      element === originalRange.startContainer ||
      element.contains(originalRange.startContainer)
    const selectionEndsWithinElement =
      element === originalRange.endContainer ||
      element.contains(originalRange.endContainer)
    selection.removeAllRanges()

    if (selectionStartsWithinElement || selectionEndsWithinElement) {
      temporaryRange.selectNodeContents(element)

      if (selectionStartsWithinElement) {
        temporaryRange.setStart(
          originalRange.startContainer,
          originalRange.startOffset,
        )
      }
      if (selectionEndsWithinElement) {
        temporaryRange.setEnd(
          originalRange.endContainer,
          originalRange.endOffset,
        )
      }

      selection.addRange(temporaryRange)
    }
  }
```
- added a few more asserts to the tests.
- types
- documentation

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Updated Type Definitions
- [x] Ready to be merged

### CODE REVIEW

**Comment by silviuaavram** [comment]
Hey @gnapse could you help me with a review here, please? Thanks!

File: README.md [hunk 1]
```diff
@@ -1420,6 +1422,71 @@ expect(deleteButton).not.toHaveDescription()
...
+
+This is useful to check if text or part of the text is selected within an
+element. The element can be either an input of type text, a textarea, or any
+other element that contains text, such as a paragraph, span, div etc.
+
+NOTE: the expected selection is a string, it does not allow to check for
+selection range indeces.
+
+```typescript
+toHaveSelection(expectedSelection: string)
```
Comment (gnapse) [inline README.md:1440]: This does not document the use case where the argument can be omitted to test if there's a selection or not. Although you reflect that in the examples below, in this very markdown file, but still, it'd be good to reflect that here somehow.

```suggestion
toHaveSelection(expectedSelection?: string)
```

**Review by gnapse — APPROVED** [review]
Thanks!

**Comment by gnapse** [comment]
@all-contributors please add @silviuaavram for code, test

**Comment by levrik** [comment]
Why is this matcher deprecated from the beginning on? I'm so confused...

**Comment by silviuaavram** [comment]
It's not deprecated, it was just added to the wrong section in the README. Let me fix that.

**Pull request merged**